src/rapids_pre_commit_hooks/copyright.py

```diff
@@ -0,0 +1,213 @@
+                            ", NVIDIA CORPORATION",
+                        )
+            else:
+                linter.add_warning((0, 0), "no copyright notice found")
+
+
+def get_target_branch(repo):
+    # Try environment
+    target_branch_name = os.getenv("GITHUB_BASE_REF")
+    if target_branch_name:
+        try:
+            return repo.heads[target_branch_name]
+        except IndexError:
+            pass
+    target_branch_name = os.getenv("TARGET_BRANCH")
+    if target_branch_name:
+        try:
+            return repo.heads[target_branch_name]
+        except IndexError:
+            pass
+    target_branch_name = os.getenv("RAPIDS_BASE_BRANCH")
+    if target_branch_name:
+        try:
+            return repo.heads[target_branch_name]
+        except IndexError:
+            pass
+
+    # Try config
+    with repo.config_reader() as r:
+        target_branch_name = r.get("rapidsai", "baseBranch", fallback=None)
+    if target_branch_name:
+        try:
+            return repo.heads[target_branch_name]
+        except IndexError:
+            pass
+
+    # Try newest branch-xx.yy
+    branches = sorted(
+        (
+            (branch, (match.group("major"), match.group("minor")))
+            for branch in repo.heads
+            if (match := BRANCH_RE.search(branch.name))
+        ),
+        key=lambda i: i[1],
+        reverse=True,
+    )
+    try:
+        return branches[0][0]
+    except IndexError:
+        pass
+
+    # Appropriate branch not found
+    return None
+
+
+def get_target_branch_upstream_commit(repo):
+    target_branch = get_target_branch(repo)
+    if target_branch is None:
+        return repo.head.commit
+
+    target_branch_upstream = target_branch.tracking_branch()
+    if target_branch_upstream:
+        return target_branch_upstream.commit
+
+    def try_get_ref(remote):
+        try:
+            return remote.refs[target_branch.name]
+        except IndexError:
+            return None
+
+    candidate_upstreams = sorted(
+        (upstream for remote in repo.remotes if (upstream := try_get_ref(remote))),
+        key=lambda upstream: upstream.commit.committed_datetime,
+        reverse=True,
+    )
+    try:
+        return candidate_upstreams[0].commit
+    except IndexError:
+        pass
+
+    return target_branch.commit
+
+
+def get_changed_files(repo, target_branch_upstream_commit):
+    changed_files = {}
+
+    diffs = target_branch_upstream_commit.diff(
+        other=None,
+        merge_base=True,
+        find_copies=True,
+        find_copies_harder=True,
+        find_renames=True,
+    )
+    for diff in diffs:
+        if diff.change_type == "A":
+            changed_files[diff.b_path] = None
+        elif diff.change_type != "D":
+            changed_files[diff.b_path] = diff.a_blob
+
+    changed_files.update({f: None for f in repo.untracked_files})
+    return changed_files
+
+
+def check_copyright():
+    repo = git.Repo()
```

Comment:
One other case to test (or just manually check) is what happens when the repo where this pre-commit hook is defined is a submodule. e.g. [spark-rapids-jni](https://github.com/NVIDIA/spark-rapids-jni) uses [cudf as a submodule](https://github.com/NVIDIA/spark-rapids-jni/tree/branch-24.02/thirdparty). I wouldn't want `git.Repo()` to observe or interfere across the submodule boundary.

Reply:
Could you give more detail on what you want to test here? Maybe write a simple test case as a shell script?